README.txt
FriedZopeBase
=============

Overview
--------

 License: ZPL
 Copyright, Peter Bengtsson, peter@fry-it.com, Fry-IT, 2007

 A set of base classes with useful methods attached for Zope2 Product
 development. 
 
 
Note
----

 This package was developed initially many years ago by Peter
 Bengtsson where he put various conventions and useful functions and
 scripts into one place. 

CHANGES.txt
- 0.22 (peter)

    Fixed all broken tests in testZope with and without slimmer
    installed.
    Vastly improved Zope.py and it's registerXXXFile functions.

- 0.21 (peter)

    Fixed important bug in sendEmailNG() that would try to send to and
    from non-ascii emails (e.g. u'test@fry-it.com')

- 0.20 (peter)

    Added ukpostcode which exposes two useful methods:
     valid_uk_postcode, format_uk_postcode
     
- 0.19 (peter)

    Improvements to sendEmailNG() with support for charset parameter
    
    Improvements to html2plaintext when there is no body tag in the
    email.
    
    Fix for getRedirectURL() where parameter 'params' wasn't reset
    between usage. Always vary of using empty dicts or lists as default
    parameter values.

- 0.18 (lukasz)

    Functions that register files are now return number of files 
    registered
    
- 0.17

    New feature: Made it possible to set environment variable
    ABSOLUTE_MISC_URLS to get http://.../misc_/... instead of just
    /misc_/...

- 0.16

    New feature: made sendEmailNG() smarter. If the 'debug' parameter
    is a valid directory or a valid file it'll write to that instead
    of stdout.

- 0.15

    Bug fixed: GzippedImageFile.index_html() had an error on 
    self.original_content_size

- 0.14

    New feature: set_cookie() accepts 'across_domain_cookie_' which sets
    the cookie acros multiple subdomains like 'm.peterbe.com' AND
    'www.peterbe.com'
    
    New feature: attachImage() and attachImages() sets imagefiles into
    classes as class attributes.
    
    New feature: Utils.iuniqify() does what uniqify does but case
    insensitively if possible.
    
    New feature: http_redirect() can sniff for getRootURL() method.
    
- 0.13
  
    Bug fixed: doCache() rounded off hours if a float < 24 wrongly.
    
- 0.12
  
    Bug fixed: Zope.registerIcons() can now accept a list of strings
    of imagenames.

    New feature: registerIcon() prevents paths with '//' in them.

- 0.11
  
    New feature: Added stopCache() and doCache() for HTTP header caching
    
- 0.10
  
    Bug fixed: sendEmail() sends the email differetly if the found
    Mailhost is a SecureMailHost or regular MailHost object so that the
    headers look right.
    
- 0.9
  
    New feature: sendEmail() takes a new parameter 'debug' which can
    be a boolean or an object that has a 'write()' method eg. StringIO
    instances or sys.stderr
    
- 0.8
  
    Bug fixed: registerIcons() accepts Globals=globals() parameter.
    
- 0.7
  
    Bug fixed: _deployImages() skips '.svn' and 'CVS' folders.

    New feature: sendEmail() uses secureSend() if mailhost is
    SecureMailHost
    
- 0.6

    New feature: sendEmail(msg, to, from, subject, swallowerrors=False)

    New feature: Added expire_cookie() method to Bases
    
- 0.5

    New feature: http_redirect() can take a dict as second param or
    applied keyword arguments and these are added to the URL.
    
- 0.4

    New feature: Base http_redirect()
    
- 0.3

    New feature: Added HomepageBTreeBase and 
    HomepageBTreeBaseCatalogAware classes.
    
- 0.2

    New feature: set_cookie() makes sure DateTime objects are
    converted to RFC822 format.

- 0.1

    Project started.

addhrefs.py
##
## addhrefs.py
## by Peter Bengtsson, 2004-2005, mail@peterbe.com
##
## License: ZPL (http://www.zope.org/Resources/ZPL)
##
__doc__='''A little function that puts HTML links into text.'''
__version__='0.7'


__changes__ = '''
0.7           addhrefs() checks that emaillinkfunction and urllinkfunction if passed
              are callable.
    
0.6           three new parameters:
                 return_everything=0 - returns (text, urlinfo, emailinfo)
				       where urlinfo and emailinfo are lists
		 emaillinkfunction=None - callback function for making an email
		                          into a HTML link the way you want it
		 urllinkfunction=None - callback function for making an URL into
					a HTML link the way you want it
					
0.5           "foo https bar" created one link

0.4           Python2.1 compatible with custom True,False

0.3           Only one replace which solved the problem with
              "www.p.com bla bla http://www.p.com"
	      
0.2           Lots of fine tuning. Created unit tests.

0.1           Project started. First working draft
'''

__credits__='''
David Otton,
"flump cakes"
'''

True = not not 1
False = not True
		     
import re, sys


_end_dropouts = list(')>.;:,"')
_start_dropouts = list('(<')
def _massageURL(url):
    while url[-1] in _end_dropouts:
        url = url[:-1]
    if url[0] in _start_dropouts:
        url = url[1:]
  
    return url

def _improveURL(url):
    if url.startswith('www.'):
        return 'http://'+url
    return url


def _makeLink(url):
    return '<a href="%s">%s</a>'%(_improveURL(url), url)

def _makeMailLink(url):
    return '<a href="mailto:%s">%s</a>'%(_improveURL(url), url)

def _rejectEmail(email, start):
    if email.find(':') > -1:
	return True
    return False

_bad_in_url = list('!()<>')
_dont_start_url = list('@')
def _rejectURL(url, start):
    """ return true if the URL can't be a URL """
    if url.lower()=='https':
	return True
    for each in _bad_in_url:
	if url.find(each) > -1:
	    return True
    whereat = url.find('@')
    if whereat > -1:
	url = url.replace('http://','').replace('ftp://','')
	if not -1 < url.find(':') < whereat:
	    return True
    if start in _dont_start_url:
	return True
    return False
    

_mailto_regex = re.compile('((^|\(|<|\s|)(\S+@\S+\.\S+)(\)|>|\s|$))')
_url_regex = re.compile('((^|\(|<|@|\s|)(ftp\S+|http\S+|www\.\S+)(\)|>|\s|$))')
def addhrefs(text, return_everything=0, 
	     emaillinkfunction=_makeMailLink,
	     urllinkfunction=_makeLink):
    
    if not callable(emaillinkfunction):
	if emaillinkfunction is not None:
	    print >>sys.stderr, "%r is not callable email link function"%emaillinkfunction
	emaillinkfunction = _makeMailLink
    
    if not callable(urllinkfunction):
	if urllinkfunction is not None:
	    print >>sys.stderr, "%r is not callable URL link function"%urllinkfunction
	urllinkfunction = _makeLink
    
    info_emails = []
    info_urls = []
    
    urls = _url_regex.findall(text)
    for each in urls:
        whole, start, url, end = each

        url = _massageURL(url)
	if _rejectURL(url, start):
	    continue
	link = urllinkfunction(url)
	if return_everything:
	    info_urls.append((url, link))
        better = whole.replace(url, link)
        text = text.replace(whole, better, 1)
        
    
    for each in _mailto_regex.findall(text):
        whole, start, url, end = each
        url = _massageURL(url)
	if _rejectEmail(url, start):
	    continue
	if url.find(':') > -1:
	    link = urllinkfunction(url)
	    if return_everything:
		info_urls.append((url, link))
	    better = whole.replace(url, link)
	else:
	    link = emaillinkfunction(url)
	    info_emails.append((url, link))
	    better = whole.replace(url, link)
        text = text.replace(whole, better)        

    if return_everything:
	return text, info_urls, info_emails
    else:
	return text


def test():
    t="this some text http://www.peterbe.com/ with links www.peterbe.com in it"
    t='''this <a href="http://www.google.com">some</a> text http://www.peterbe.com/
    with links www.peterbe.com in it <a href="http://www.example.com">Example</a>'''
    

    

    t2='this <a href="http://www.google.com">some</a> text http://www.peterbe.com/ '\
       'with links www.peterbe.com in it '\
      '<a href="http://www.example.com">Example</a>'

    t3='''this <a href="http://www.google.com">some</a> text http://www.peterbe.com/
with links www.peterbe.com in it <a href="http://www.example.com">Example</a>
www,peterbe.com and <a href="www.peterbe.com">www.peterbe.com</a>
    '''

    t4='''https://www.imdb.com (www.peterbe.com/?a=e) asd tra la www.google.com'''
    
    t = 'word (www.peterbe.com) word'
    
    t = 'word <http://www.peterbe.com?poage=2314> and so on'
    
    t = 'Go to: http://www.peterbe.com. There youll find'
    
    t = 'Go to: http://www.peterbe.com:'
    t = '''https://www.imdb.com.'''
    t = 'Hello mail@peterbe.com to you'
    t = 'Hello <mail@peterbe.com> and <www.google.com> to you'
    #t = open('sample-htmlfree.txt').read()
    t = '<a href="link1.html">Link1</a> link www.2.com'
    t = "<a href='www.google.com'>Link1</a> link www.2.com"
    t = '''1. http://www.peterbe.com
2. www.peterbe.com
3. <http://www.peterbe.com>
4. mail@foobar.com
5. "Name <mail@foobar.com>"'''
    t = 'xxx mail@peterbe.com peter@grenna.net'
    t += ' xxx www.peterbe.com www.google.com xxx'

    t = 'mail@peterbe.com 123@a.com or www2.ibm.com or www.ibm.com?asda=ewr&amp;gr:int=34.'
    
    t = 'peter@grenna.net 123@a.com ftp://ftp.uk.linux.org/'
    t = 'http://david:otton@www.something.com david:otton@www.something.com'
    t = ''' <http://www.google.com/?asd=q32> xxx
    <http://www.peterbe.com> abc <mail@peterbe.com>'''
    t='''www.msn.co.uk

http://msn.co.uk
http://www.msn.co.uk

ftp:/google.com
'''
    t = 'At http://localhost/ I have apache and at http://localhost:8080 '\
        ' I have Zope David used http://enchanter or http://enchanter/'
    t = 'See http://www.something.com/page?this=that#001'
    t = 'Bla bla https bla bla and http bla'
    print addhrefs(t, emaillinkfunction=None)
    


        
    
    
if __name__=='__main__':
    test()
